File: src/main/java/me/dkim19375/unig0/util/FileUtils.java
package me.dkim19375.unig0.util;

import me.dkim19375.unig0.UniG0;
import me.dkim19375.unig0.util.properties.GlobalProperties;
import me.dkim19375.unig0.util.properties.ServerProperties;
import me.mattstudios.config.SettingsManager;

import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class FileUtils {
    public static void reload() {
        UniG0.getFileManager().getGlobalConfig().reload();
        Map<String, SettingsManager> map = UniG0.getFileManager().getServerConfigs();
        for (SettingsManager manager : map.values()) {
            manager.reload();
        }
    }

    public static void save() {
        UniG0.getFileManager().getGlobalConfig().save();
        Map<String, SettingsManager> map = UniG0.getFileManager().getServerConfigs();
        for (SettingsManager manager : map.values()) {
            manager.save();
        }
    }

    public static void reset() {
        Map<String, SettingsManager> map = UniG0.getFileManager().getServerConfigs();
        for (SettingsManager manager : map.values()) {
            manager.set(ServerProperties.prefix, "?");
            manager.set(ServerProperties.delete_commands, new HashSet<>());
            manager.set(ServerProperties.disabled_channels, new HashSet<>());
            manager.set(ServerProperties.welcomer_dm, "Welcome to {ServerName}!\nHave a great time!");
            manager.set(ServerProperties.welcomer_message, "{user} has join the server!\nWe are now at {members} members!");
            manager.set(ServerProperties.welcomer_channel, "");
            manager.set(ServerProperties.welcomer_enabled_dm, true);
            manager.set(ServerProperties.welcomer_enabled_message, true);
        }
        save();
    }

    public static String getPrefix(String id) {
        return UniG0.getFileManager().getServerConfig(id).get(ServerProperties.prefix);
    }

    public static void setPrefix(String id, String prefix) {
        UniG0.getFileManager().getServerConfig(id).set(ServerProperties.prefix, prefix);
        save();
    }

    public static String getToken() {
        return UniG0.getFileManager().getGlobalConfig().get(GlobalProperties.token);
    }

    public static Set<String> getDeletedCommands(String id) {
        Set<String> before = new HashSet<>(UniG0.getFileManager().getServerConfig(id).get(ServerProperties.delete_commands));
        if (before.contains("*")) {
            Set<String> set = new HashSet<>();
            set.add("*");
            return set;
        }
        return before;
    }

    public static boolean shouldDeleteCommand(String id, String command) {
        return UniG0.getFileManager().getServerConfig(id).get(ServerProperties.delete_commands).contains(command);
    }

    public static void setDeletedCommands(String id, Set<String> commands) {
        UniG0.getFileManager().getServerConfig(id).set(ServerProperties.delete_commands, commands);
        save();
    }

    public static void removeDeletedCommand(String id, String command) {
        Set<String> set = new HashSet<>(getDeletedCommands(id));
        set.remove(command);
        UniG0.getFileManager().getServerConfig(id).set(ServerProperties.delete_commands, set);
        save();
    }

    public static void addDeletedCommand(String id, String command) {
        Set<String> set = new HashSet<>(getDeletedCommands(id));
        set.add(command);
        UniG0.getFileManager().getServerConfig(id).set(ServerProperties.delete_commands, set);
        save();
    }

    public static Set<String> getDisabledChannels(String id) {
        return UniG0.getFileManager().getServerConfig(id).get(ServerProperties.disabled_channels);
    }

    public static boolean shouldDisableChannel(String id, String channelID) {
        return UniG0.getFileManager().getServerConfig(id).get(ServerProperties.disabled_channels).contains(channelID);
    }

    public static void setDisabledChannels(String id, Set<String> channelID) {
        UniG0.getFileManager().getServerConfig(id).set(ServerProperties.disabled_channels, channelID);
        save();
    }

    public static void removeDisabledChannel(String id, String channelID) {
        Set<String> set = new HashSet<>(getDisabledChannels(id));
        set.remove(channelID);
        UniG0.getFileManager().getServerConfig(id).set(ServerProperties.disabled_channels, set);
        save();
    }

    public static void addDisabledChannel(String id, String channelID) {
        Set<String> set = new HashSet<>(getDisabledChannels(id));
        set.add(channelID);
        UniG0.getFileManager().getServerConfig(id).set(ServerProperties.disabled_channels, set);
        save();
    }

    public static boolean isWelcomeMessageEnabled(String id) {
        return UniG0.getFileManager().getServerConfig(id).get(ServerProperties.welcomer_enabled_message);
    }

    public static void setWelcomeMessageEnabled(String id, boolean enabled) {
        UniG0.getFileManager().getServerConfig(id).set(ServerProperties.welcomer_enabled_message, enabled);
        save();
    }

    public static boolean isWelcomeDMEnabled(String id) {
        return UniG0.getFileManager().getServerConfig(id).get(ServerProperties.welcomer_enabled_dm);
    }

    public static void setWelcomeDMEnabled(String id, boolean enabled) {
        UniG0.getFileManager().getServerConfig(id).set(ServerProperties.welcomer_enabled_dm, enabled);
        save();
    }

    public static String getWelcomeMessage(String id) {
        return UniG0.getFileManager().getServerConfig(id).get(ServerProperties.welcomer_message);
    }

    public static void setWelcomeMessage(String id, String message) {
        UniG0.getFileManager().getServerConfig(id).set(ServerProperties.welcomer_message, message);
        save();
    }

    public static String getDMMessage(String id) {
        return UniG0.getFileManager().getServerConfig(id).get(ServerProperties.welcomer_dm);
    }

    public static void setDMMessage(String id, String message) {
        UniG0.getFileManager().getServerConfig(id).set(ServerProperties.welcomer_dm, message);
        save();
    }

    public static String getWelcomeChannel(String id) {
        return UniG0.getFileManager().getServerConfig(id).get(ServerProperties.welcomer_channel);
    }

    public static void setWelcomeChannel(String id, String channelId) {
        UniG0.getFileManager().getServerConfig(id).set(ServerProperties.welcomer_channel, channelId);
        save();
    }
}

File: src/main/kotlin/me/dkim19375/unig0/util/FileUtils.kt
package me.dkim19375.unig0.util

import me.dkim19375.dkim19375jdautils.util.getIgnoreCase
import me.dkim19375.unig0.UniG0
import me.dkim19375.unig0.util.property.GlobalProperties
import me.dkim19375.unig0.util.property.ServerProperties

@Suppress("unused")
object FileUtils {
    fun reload(bot: UniG0) {
        bot.fileManager.globalConfig.reload()
        val map = bot.fileManager.getServerConfigs()
        for (manager in map.values) {
            manager.reload()
        }
    }

    @Suppress("MemberVisibilityCanBePrivate")
    fun save(bot: UniG0) {
        bot.fileManager.globalConfig.save()
        val map = bot.fileManager.getServerConfigs()
        for (manager in map.values) {
            manager.save()
        }
    }

    fun reset(bot: UniG0) {
        val map = bot.fileManager.getServerConfigs()
        for (manager in map.values) {
            manager.set(ServerProperties.prefix, "?")
            manager.set(ServerProperties.delete_commands, emptySet())
            manager.set(ServerProperties.disabled_channels, emptySet())
            manager.set(ServerProperties.welcomer_dm, "Welcome to {ServerName}!\nHave a great time!")
            manager.set(
                ServerProperties.welcomer_message,
                "{user} has join the server!\nWe are now at {members} members!"
            )
            manager.set(ServerProperties.welcomer_channel, "")
            manager.set(ServerProperties.welcomer_enabled_dm, true)
            manager.set(ServerProperties.welcomer_enabled_message, true)
        }
        save(bot)
    }

    fun getPrefix(id: String, bot: UniG0): String {
        return bot.fileManager.getServerConfig(id).get(ServerProperties.prefix)
    }

    fun setPrefix(id: String, prefix: String, bot: UniG0) {
        bot.fileManager.getServerConfig(id).set(ServerProperties.prefix, prefix)
        save(bot)
    }

    fun getToken(bot: UniG0): String = bot.fileManager.globalConfig.get(GlobalProperties.token)

    fun getDisabledCommands(id: String, bot: UniG0): Set<String> = bot.fileManager.getServerConfig(id).get(ServerProperties.disabled_commands).toSet()

    fun getDeletedCommands(id: String, bot: UniG0): Set<String> {
        val before = bot.fileManager.getServerConfig(id).get(ServerProperties.delete_commands).toSet()
        if (before.contains("*")) {
            val set = mutableSetOf<String>()
            set.add("*")
            return set
        }
        return before
    }

    fun shouldDeleteCommand(id: String, command: String, bot: UniG0): Boolean {
        return bot.fileManager.getServerConfig(id).get(ServerProperties.delete_commands).contains(command)
    }

    fun setDeletedCommands(id: String, commands: Set<String>, bot: UniG0) {
        bot.fileManager.getServerConfig(id).set(ServerProperties.delete_commands, commands)
        save(bot)
    }

    fun removeDeletedCommand(id: String, command: String, bot: UniG0) {
        val set = getDeletedCommands(id, bot).toMutableSet()
        set.remove(command)
        bot.fileManager.getServerConfig(id).set(ServerProperties.delete_commands, set)
        save(bot)
    }

    fun addDeletedCommand(id: String, command: String, bot: UniG0) {
        val set = getDeletedCommands(id, bot).toMutableSet()
        set.add(command)
        bot.fileManager.getServerConfig(id).set(ServerProperties.delete_commands, set)
        save(bot)
    }

    fun enableCommand(id: String, command: String, bot: UniG0) {
        val set = getDisabledCommands(id, bot).toMutableSet()
        set.remove(set.getIgnoreCase(command) ?: "")
        bot.fileManager.getServerConfig(id).set(ServerProperties.disabled_commands, set)
        save(bot)
    }

    fun disableCommand(id: String, command: String, bot: UniG0) {
        val set = getDisabledCommands(id, bot).toMutableSet()
        set.remove(set.getIgnoreCase(command) ?: "")
        set.add(command)
        bot.fileManager.getServerConfig(id).set(ServerProperties.disabled_commands, set)
        save(bot)
    }

    fun getDisabledChannels(id: String, bot: UniG0): Set<String> {
        return bot.fileManager.getServerConfig(id).get(ServerProperties.disabled_channels)
    }

    fun shouldDisableChannel(id: String, channelID: String, bot: UniG0): Boolean {
        return bot.fileManager.getServerConfig(id).get(ServerProperties.disabled_channels).contains(channelID)
    }

    fun setDisabledChannels(id: String, channelID: Set<String>, bot: UniG0) {
        bot.fileManager.getServerConfig(id).set(ServerProperties.disabled_channels, channelID)
        save(bot)
    }

    fun removeDisabledChannel(id: String, channelID: String, bot: UniG0) {
        val set = getDisabledChannels(id, bot).toMutableSet()
        set.remove(channelID)
        bot.fileManager.getServerConfig(id).set(ServerProperties.disabled_channels, set)
        save(bot)
    }

    fun addDisabledChannel(id: String, channelID: String, bot: UniG0) {
        val set = getDisabledChannels(id, bot).toMutableSet()
        set.add(channelID)
        bot.fileManager.getServerConfig(id).set(ServerProperties.disabled_channels, set)
        save(bot)
    }

    fun isWelcomeMessageEnabled(id: String, bot: UniG0): Boolean {
        return bot.fileManager.getServerConfig(id).get(ServerProperties.welcomer_enabled_message)
    }

    fun setWelcomeMessageEnabled(id: String, enabled: Boolean, bot: UniG0) {
        bot.fileManager.getServerConfig(id).set(ServerProperties.welcomer_enabled_message, enabled)
        save(bot)
    }

    fun isWelcomeDMEnabled(id: String, bot: UniG0): Boolean {
        return bot.fileManager.getServerConfig(id).get(ServerProperties.welcomer_enabled_dm)
    }

    fun setWelcomeDMEnabled(id: String, enabled: Boolean, bot: UniG0) {
        bot.fileManager.getServerConfig(id).set(ServerProperties.welcomer_enabled_dm, enabled)
        save(bot)
    }

    fun getWelcomeMessage(id: String, bot: UniG0): String {
        return bot.fileManager.getServerConfig(id).get(ServerProperties.welcomer_message)
    }

    fun setWelcomeMessage(id: String, message: String, bot: UniG0) {
        bot.fileManager.getServerConfig(id).set(ServerProperties.welcomer_message, message)
        save(bot)
    }

    fun getDMMessage(id: String, bot: UniG0): String {
        return bot.fileManager.getServerConfig(id).get(ServerProperties.welcomer_dm)
    }

    fun setDMMessage(id: String, message: String, bot: UniG0) {
        bot.fileManager.getServerConfig(id).set(ServerProperties.welcomer_dm, message)
        save(bot)
    }

    fun getWelcomeChannel(id: String, bot: UniG0): String {
        return bot.fileManager.getServerConfig(id).get(ServerProperties.welcomer_channel)
    }

    fun setWelcomeChannel(id: String, channelId: String, bot: UniG0) {
        bot.fileManager.getServerConfig(id).set(ServerProperties.welcomer_channel, channelId)
        save(bot)
    }
}